File: src/main/java/io/github/Earth1283/fixLag/managers/ExperienceOrbMerger.java
package io.github.Earth1283.fixLag.managers;

import io.github.Earth1283.fixLag.FixLag;
import org.bukkit.Bukkit;
import org.bukkit.World;
import org.bukkit.entity.Entity;
import org.bukkit.entity.ExperienceOrb;
import org.bukkit.scheduler.BukkitRunnable;

import java.util.Collection;
import java.util.HashSet;
import java.util.Set;

public class ExperienceOrbMerger extends BukkitRunnable {

    private final FixLag plugin;
    private final ConfigManager config;

    public ExperienceOrbMerger(FixLag plugin) {
        this.plugin = plugin;
        this.config = plugin.getConfigManager();
    }

    @Override
    public void run() {
        if (!config.isXpOrbMergerEnabled()) return;

        double radius = config.getXpOrbMergerRadius();
        Set<ExperienceOrb> processed = new HashSet<>();

        for (World world : Bukkit.getWorlds()) {
            for (ExperienceOrb orb : world.getEntitiesByClass(ExperienceOrb.class)) {
                if (processed.contains(orb) || orb.isDead()) continue;

                Collection<Entity> nearby = world.getNearbyEntities(orb.getLocation(), radius, radius, radius, e -> e instanceof ExperienceOrb && e != orb && !e.isDead() && !processed.contains(e));
                
                if (!nearby.isEmpty()) {
                    int totalXp = orb.getExperience();
                    for (Entity e : nearby) {
                        ExperienceOrb other = (ExperienceOrb) e;
                        totalXp += other.getExperience();
                        other.remove();
                        processed.add(other);
                    }
                    orb.setExperience(totalXp);
                }
                processed.add(orb);
            }
        }
    }
}

File: src/main/kotlin/io/github/Earth1283/fixLag/managers/ExperienceOrbMerger.kt
package io.github.Earth1283.fixLag.managers

import io.github.Earth1283.fixLag.FixLag
import org.bukkit.Bukkit
import org.bukkit.entity.ExperienceOrb
import org.bukkit.scheduler.BukkitRunnable

class ExperienceOrbMerger(private val plugin: FixLag) : BukkitRunnable() {

    private val config: ConfigManager = plugin.configManager

    override fun run() {
        if (!config.isXpOrbMergerEnabled) return

        val radius = config.xpOrbMergerRadius
        val processed = mutableSetOf<ExperienceOrb>()

        for (world in Bukkit.getWorlds()) {
            for (orb in world.getEntitiesByClass(ExperienceOrb::class.java)) {
                if (orb in processed || orb.isDead) continue

                val nearby = world.getNearbyEntities(orb.location, radius, radius, radius) { e ->
                    e is ExperienceOrb && e !== orb && !e.isDead && e !in processed
                }

                if (nearby.isNotEmpty()) {
                    var totalXp = orb.experience
                    for (e in nearby) {
                        val other = e as ExperienceOrb
                        totalXp += other.experience
                        other.remove()
                        processed.add(other)
                    }
                    orb.experience = totalXp
                }
                processed.add(orb)
            }
        }
    }
}
